### PR TITLE
Simple terraform provider CICD module

### PR DESCRIPTION
**Pull request by mbonig**
In a comment on the pull request, include the following information and check off the list.

 - [x] Mini bio
  - Soft capped to 75 words
 - [x] Desired date of publishing
  - Make sure this matches the `date` in the front matter
 - [x] Social media contact info
 - [x] Redaction: (yes|no)
  - Allow redaction, editing or structure changes by technical writer staff
 - [x] Comments: (yes|no)
  - If you're willing to answer comments about your post, set `comments: true`.
  - Make sure to subscribe to your post comments by clicking on the **Subscribe** link at the very bottom of the post _once_ it's published.

@catlook Please review.

### CODE REVIEW

**Comment by rackernexus** [comment]
Can one of the admins verify this patch?

**Comment by rackernexus** [comment]
:eyes: [Preview](https://staging.developer.rackspace.com/build-24a433c927/)

File: _posts/2019-02-14-simple-terraform-provider-cicd.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  - Automation
+---
+
+Terraform has gained a lot of popularity in the last couple years. Rackspace prefers to use Terraform to quickly spin up new architecture in AWS and Azure. However, with Amazon's lightning-fast deployment of new features, it has become harder for the Provider maintainers to keep up. Developers are left waiting for new features to be developed and merged into the `master` branch before becoming available for general consumption.
+
+<!-- more -->
+
+Open Source to the rescue! Since the source for Terraform Providers are available via Github, any new feature doesn't have to be in the `master` (mainline) Provider before it can be leveraged by Terraform users.
+
+This post will discuss a simple Terraform module that was developed to allow anyone to easily clone, build, and install a fork or branch of a Terraform Provider without having to setup a GOLANG build environment directly.
```
Comment (catlook) [inline _posts/2019-02-14-simple-terraform-provider-cicd.md]: This post discusses a ...

File: _posts/2019-02-14-simple-terraform-provider-cicd.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+### The Need
+
+Currently (Jan 2019) there are almost 2000 open issues with the [Terraform AWS Provider](https://github.com/terraform-providers/terraform-provider-aws/issues). That's a lot of issues... and while Hashicorp works to make Terraform 0.12 available, it feels that some of this development work to get new features and bug fixes merged into `master` has slowed down. Thankfully the robust community has - in many cases - developed solutions and submitted PRs. Currently (Jan 2019) there are 267 outstanding PRs. That's a lot of PRs...
```
Comment (catlook) [inline _posts/2019-02-14-simple-terraform-provider-cicd.md]: Currently (Jan, 2019), there ...
...  down. Thankfully, the ...
... PRs. As of January, 2019, there

File: _posts/2019-02-14-simple-terraform-provider-cicd.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+Let's take a look at one specific issue, the [Amazon FSx Feature](https://github.com/terraform-providers/terraform-provider-aws/issues/7035). FSx was recently announced (Dec 2019) and is a drop-in replacement for Windows File Server as a managed service (see [this blog](https://developer.rackspace.com/blog/amazon-fsx-replaces-file-shares/)). Any team wanting to deploy FSx would need to find some other method for deploying it. With this PR unmerged, we'd need to either fallback to CLI commands, or CloudFormation (which at the time of this writing is still not an option as even AWS has not released the CloudFormation changes required to support FSx), or manage this through the Console. None of these options are good for us.
```
Comment (catlook) [inline _posts/2019-02-14-simple-terraform-provider-cicd.md]: ... (Dec, 2019) ...

File: _posts/2019-02-14-simple-terraform-provider-cicd.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+However, since a nice user in the community created a PR, we can review that and decide if we want to use that code instead.
+
+### The Problem
+
+While this sounds simple, it isn't always easy. All Terraform Providers are written in GOLANG, so cloning that fork, building, and deploying that to our environments to use instead of the mainline `terraform-provider-aws`, is not trivial.
```
Comment (catlook) [inline _posts/2019-02-14-simple-terraform-provider-cicd.md]: ... deploying a fix or enhancement to our ...

File: _posts/2019-02-14-simple-terraform-provider-cicd.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+Let's assume you've done Go development before, then you pull the code, build it, run tests, then copy the resulting file to your `~/.terraform.d/plugins` directory, no problem.
```
Comment (catlook) [inline _posts/2019-02-14-simple-terraform-provider-cicd.md]: ... before. In that case, you pull ...

File: _posts/2019-02-14-simple-terraform-provider-cicd.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+But, if you're like me and had never done Go, are you willing to spend the time to get things setup properly? Here's a tip that took me a couple of hours to figure out: GOLANG cares about your directory structure!
```
Comment (catlook) [inline _posts/2019-02-14-simple-terraform-provider-cicd.md]: ... things set up properly? ...

File: _posts/2019-02-14-simple-terraform-provider-cicd.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+Either way, at the end I felt things could be easier.
```
Comment (catlook) [inline _posts/2019-02-14-simple-terraform-provider-cicd.md]: ... the end, I felt ...

File: _posts/2019-02-14-simple-terraform-provider-cicd.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  
+### The Solution
+
+So, [here](https://github.com/rackerlabs/simple-provider-cicd) is a Terraform Module, Open Source and available on Github, that you can use to setup a CICD process using CodePipeline and CodeBuild to build and deploy any fork or branch of a Terraform Provider to an S3 bucket for consumption. Here's how it works:
```
Comment (catlook) [inline _posts/2019-02-14-simple-terraform-provider-cicd.md]: ... use to set up a CICD ...

File: _posts/2019-02-14-simple-terraform-provider-cicd.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+1. Execute the TF Module, providing the Github owner, repository, and branch to build, along with a couple of additional parameters. This creates all the necessary setup.
+
+Then, on any push:
```
Comment (catlook) [inline _posts/2019-02-14-simple-terraform-provider-cicd.md]: add a blank line after this one, so numbering works.

File: _posts/2019-02-14-simple-terraform-provider-cicd.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+2. CodePipeline pulls source from Github.
+3. CodeBuild builds that source into a `darwin_amd64` and `linux_amd64` binaries.
+4. CodePipeline copies those two binaries to an S3 bucket for consumption (this is private, btw!).
+5. Any user who wants to use this provider just copies the resulting binary to their own machine. Getting it out of that S3 bucket is on you, though.
+ 
+Please review the README and feel free to open any issues or provide feedback.
+
+### An Example
+
+Let's take a look at an example of using this with the FSx PR I mentioned earlier. First, I would create a new Terraform module and include a main.tf:
```
Comment (catlook) [inline _posts/2019-02-14-simple-terraform-provider-cicd.md]: ... PR that I mentioned ...

File: _posts/2019-02-14-simple-terraform-provider-cicd.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    github__repo   = "terraform-provider-aws"
+    github__branch = "resource-fsx-file-system"
+    project_name   = "fsx-aws-provider"
+    s3_bucket      = "terraform-provider-aws-fsx"
+    go_package     = "github.com/terraform-providers/terraform-provider-aws"  
+}
+
+```
+
+Now just run your Terraform init/apply (don't forget to set a GITHUB_TOKEN first!):
```
Comment (catlook) [inline _posts/2019-02-14-simple-terraform-provider-cicd.md]: add a blank line after this one so following code block works

File: _posts/2019-02-14-simple-terraform-provider-cicd.md (hunk 9 above)
Comment (mbonig) [reply _posts/2019-02-14-simple-terraform-provider-cicd.md]: Slightly changed this already.

**Comment by rackernexus** [comment]
:mag_right: [Preview](https://staging.developer.rackspace.com/build-f7d116553a/)

**Pull request merged**